kernel/aster-nix/src/vm/vmar/vm_mapping.rs

```diff
@@ -194,22 +196,41 @@ impl VmMapping {
         let write_perms = VmPerms::WRITE;
         self.check_perms(&write_perms)?;
 
-        let mut page_addr =
-            self.map_to_addr() - self.vmo_offset() + page_idx_range.start * PAGE_SIZE;
-        for page_idx in page_idx_range {
-            let parent = self.parent.upgrade().unwrap();
-            let vm_space = parent.vm_space();
-
-            // The `VmMapping` has the write permission but the corresponding PTE is present and is read-only.
-            // This means this PTE is set to read-only due to the COW mechanism. In this situation we need to trigger a
-            // page fault before writing at the VMO to guarantee the consistency between VMO and the page table.
-            let need_page_fault = vm_space
-                .query(page_addr)?
-                .is_some_and(|prop| !prop.flags.contains(PageFlags::W));
-            if need_page_fault {
-                self.handle_page_fault(page_addr, false, true)?;
+        // We need to make sure the mapping exists.
+        //
+        // Also, if the `VmMapping` has the write permission but the corresponding
+        // PTE is present and is read-only, it would be a copy-on-write page. In
+        // this situation we need to trigger a page fault before writing at the
+        // VMO to guarantee the consistency between VMO and the page table.
+        {
+            let virt_addr =
+                self.map_to_addr() - self.vmo_offset() + page_idx_range.start * PAGE_SIZE;
+            let virt_range = virt_addr..virt_addr + page_idx_range.len() * PAGE_SIZE;
+
+            // FIXME: any sane developer would recommend using `parent.vm_space().cursor(&virt_range)`
+            // to lock the range and check the mapping status. However, this will cause a deadlock because
+            // `Self::handle_page_fault` would like to create a cursor again. The following implementation
+            // indeed introduces a TOCTOU bug.
+            for page_va in virt_range.step_by(PAGE_SIZE) {
+                let parent = self.parent.upgrade().unwrap();
+                let mut cursor = parent
+                    .vm_space()
+                    .cursor(&(page_va..page_va + PAGE_SIZE))
+                    .unwrap();
+                let map_info = cursor.query().unwrap();
+                drop(cursor);
```

Comment:
I do not strongly prefer this API design as it may make it over-complicated and hard to document. However it is indeed better to describe the locking behavior in the doc.